rpc/src/responses/json_array.rs

```diff
@@ -1,47 +1,280 @@
+            mock_value_arr.iter().zip(json_array.into_iter());
+
+        for tuple in mock_value_and_parsed_value_tuple_iter {
+            let mock_val = tuple.0;
+            let parsed_val = &tuple.1;
+            assert_eq!(mock_val, parsed_val);
+        }
+    }
+
+    #[test]
+    fn two_dimensional_string_json_array_from_nested_str_parse_ok() {
+        let mock_values = [["foo", "bar"], ["foo_", "bar_"]];
+        let mock_str_to_parse = format!("{:?}", &mock_values);
+
+        type NestedType = JsonArray<String>;
+        let parse_response = JsonArray::<NestedType>::from_nested_response_str(&mock_str_to_parse);
+        assert!(parse_response.is_ok());
+
+        let response = parse_response.unwrap();
+        let tuple_vec = get_tuple_vec_from_response_and_mock_values(response, &mock_values);
+
+        assert!(check_tuples_are_eq(tuple_vec));
+    }
+
+    #[test]
+    fn two_dimensional_unistring_json_array_from_nested_str_parse_ok() {
+        let mock_values = [["foo", "bar"], ["foo_", "bar_"]];
+        let mock_str_to_parse = format!("{:?}", &mock_values);
+
+        type NestedType = JsonArray<Unistring>;
+        let parse_response = JsonArray::<NestedType>::from_nested_response_str(&mock_str_to_parse);
+        assert!(parse_response.is_ok());
+
+        let response = parse_response.unwrap();
+        let tuple_vec = get_tuple_vec_from_response_and_mock_values(response, &mock_values);
+
+        assert!(check_tuples_are_eq(tuple_vec));
+    }
+
+    #[test]
+    fn one_dimensional_json_array_of_strings_from_str_parse_ok() {
+        let mock_values = ["foo", "bar"];
+        let mock_str_to_parse = format!("{:?}", &mock_values);
+
+        let parse_response = JsonArray::<String>::from_response_str(&mock_str_to_parse);
+        assert!(parse_response.is_ok());
+
+        let json_array = parse_response.unwrap();
+        let string_to_compare = format!("{}, {}", mock_values[0], mock_values[1]);
+        assert_eq!(json_array.to_string(), string_to_compare);
+
+        assert_eq!(json_array.into_vec().len(), mock_values.len());
+    }
+
+    #[test]
+    fn no_json_array_values_parse_ok() {
+        let mock_str_to_parse = "[]";
+
+        let parse_response = JsonArray::<String>::from_response_str(&mock_str_to_parse);
+        assert!(parse_response.is_ok());
+
+        let json_array = parse_response.unwrap().into_vec();
+        assert_eq!(json_array.len(), 0);
+    }
+
+    #[test]
+    fn empty_json_str_fail() {
+        let mock_str_to_parse = "";
+
+        let parse_response = JsonArray::<String>::from_response_str(&mock_str_to_parse);
+        assert!(parse_response.is_err());
+    }
+
+    fn get_tuple_vec_from_response_and_mock_values<'a, T, I: 'a>(
+        response: JsonArray<JsonArray<T>>,
+        mock_values: I,
+    ) -> Vec<(T, <<I as IntoIterator>::Item as IntoIterator>::Item)>
+    where
+        T: de::DeserializeOwned,
+        I: iter::IntoIterator,
+        I::Item: iter::IntoIterator,
```

Comment:
Was this necessary? no
Was this overkill? yes
Was this really fun to write? absolutely